File: PAT_Basic_Level_by_lua/1114 全素日/1114 全素日.lua
-- 题目编号: 1114 全素日
-- 实现原理: 对输入数字字符串的每个后缀子串，依次判断其是否为素数（从2到根号试除），若所有后缀均为素数则输出"All Prime!"。

local s = io.read("l")  -- 读取输入的数字字符串

-- 判断一个数是否为素数
local function p(x)
    if x < 2 then return false end  -- 小于2的数不是素数
    -- 从2到根号x试除
    for i = 2, math.floor(math.sqrt(x)) do
        if x % i == 0 then return false end  -- 能被整除则不是素数
    end
    return true
end

local ok = true  -- 标记是否所有后缀都是素数

-- 遍历每个后缀子串
for i = 1, #s do
    local t = s:sub(i)  -- 从第i位开始的后缀子串
    local yes = p(tonumber(t))  -- 判断是否为素数
    -- 输出当前后缀和判断结果
    print(t .. " " .. (yes and "Yes" or "No"))
    -- 如果有一个不是素数，标记ok为false
    if not yes then ok = false end
end

-- 如果所有后缀都是素数，输出提示
if ok then
    print("All Prime!")
end

File: PAT_Basic_Level_by_r/1114 全素日/1114 全素日.R
# 题目名称：全素日
# 题目编号：PAT Basic 1114
# 实现原理：
#   判断给定日期(yyyymmdd格式)是否为全素日。全素日定义：日期本身是素数，
#   且其任何以末尾数字结尾的子串对应的数字都是素数。
#   解题思路：1. 实现素数判断函数（试除法，优化：先判偶，再试奇数因子到√n）；
#   2. 从第1位到最后一位，依次取子串s[i..end]，转换为整数判断是否为素数；
#   3. 每行输出子串和Yes/No；4. 若所有子串都是素数，最后输出"All Prime!"。
#   注意：子串可能有前导零，但转成数值后不影响素数判断。
# 时间复杂度：O(L × √V)，L为日期长度(8)，V为子串最大值(约10^8)，每次判断素数到√V
# 空间复杂度：O(1)，仅使用常数空间
s <- scan("stdin", what="", quiet=TRUE)[1]
prime <- function(n) { if(n<2) return(FALSE); if(n==2) return(TRUE); if(n%%2==0) return(FALSE); if(n<9) return(TRUE); !any(n %% seq(3, floor(sqrt(n)), by=2) == 0) }
ok <- TRUE
for(i in seq_len(nchar(s))) { v <- substr(s,i,nchar(s)); yes <- prime(as.numeric(v)); cat(v, if(yes) "Yes" else "No", "\n"); ok <- ok && yes }
if(ok) cat("All Prime!\n")
